File: src/SkyBlock/enchants/sword/Lifesteal.php
<?php


namespace SkyBlock\enchants\sword;


use pocketmine\entity\effect\VanillaEffects;
use pocketmine\event\entity\EntityDamageEvent;
use pocketmine\player\Player;
use SkyBlock\enchants\armor\LifeShield;
use SkyBlock\enchants\BaseEnchantment;

class Lifesteal extends BaseMeleeEnchant {

    static int $id = 100;

    public function isApplicableTo(Player $holder) : bool {
        $heldItem = $holder->getInventory()->getItemInHand();
        $ench = $heldItem->getEnchantment(BaseEnchantment::getEnchantment(self::$id));
        if ($ench !== null) {
            $level = $ench->getLevel();
            if ($level === 15) {
                return mt_rand(1, 20) === 1;
            } elseif ($level === 14) {
                return mt_rand(1, 21) === 1;
            } elseif ($level === 13) {
                return mt_rand(1, 22) === 1;
            } elseif ($level === 12) {
                return mt_rand(1, 23) === 1;
            } elseif ($level === 11) {
                return mt_rand(1, 24) === 1;
            } else {
                return mt_rand(1, 25) === 1;
            }
        }
        return false;
    }

    /**
     * @param Player            $player
     * @param Player            $attacker
     * @param EntityDamageEvent $ev
     * @param int               $enchantmentlevel
     */
    public function onActivation(Player $player, Player $attacker, EntityDamageEvent $ev, int $enchantmentlevel) : void {
        $this->sendActivation($player, "§bLifeSteal §aActivated!");
        $level = BaseEnchantment::getEnchantmentLevel($attacker->getArmorInventory()->getChestplate(), LifeShield::$id);
        if ($level > 10 ? mt_rand(1, 3) === 1 : mt_rand(1, 6) === 1) {
            $this->sendActivation($player, "§bLifeSteal §cDeactivated by LifeShield Enchant!");
            return;
        }

        $attacker->getEffects()->remove(VanillaEffects::REGENERATION());
        $attacker->getEffects()->remove(VanillaEffects::HEALTH_BOOST());
        $this->sendActivation($attacker, "§cStruck by §bLifeSteal §cEnchant! Health Effects removed");

        $health2 = $attacker->getHealth();
        $health1 = $player->getHealth();
        $healthTaken = $enchantmentlevel > 9 ? 8 : ceil($enchantmentlevel / 2);

        $attacker->setHealth($health2 - $healthTaken);
        $player->setHealth($health1 + $healthTaken);

        //        if ($health2 >= 20) {
        //            $attacker->setHealth(19);
        //        } elseif ($health2 > 15) {
        //            $attacker->setHealth(14);
        //        } elseif ($health2 < 12 && $health2 > 6) {
        //            $attacker->setHealth(5);
        //        } elseif ($health2 < 3 && $health2 > 1) {
        //            $attacker->kill();
        //        }
        //
        //        if ($health1 >= 15) {
        //            $player->setHealth($health1 + 2);
        //        } elseif ($health1 < 12 && $health1 > 7) {
        //            $player->setHealth(14);
        //        } elseif ($health1 < 4 && $health1 > 1) {
        //            $player->setHealth(7);
        //        }
    }

}

File: SkyblockCore/src/main/java/dev/skyblock/enchants/sword/Lifesteal.java
package dev.skyblock.enchants.sword;
import dev.skyblock.SkyblockCore;
import dev.skyblock.enchants.BaseMeleeEnchant;
import org.bukkit.NamespacedKey;
import org.bukkit.attribute.Attribute;
import org.bukkit.entity.Player;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.potion.PotionEffectType;
import java.util.concurrent.ThreadLocalRandom;
public class Lifesteal extends BaseMeleeEnchant {
    public static final int ID = 100;
    public static final int LIFESHIELD_ID = 186;
    public Lifesteal() { super(ID, "Lifesteal", 15); }
    @Override
    public boolean isApplicableTo(Player holder, int level) {
        int bound = level >= 15 ? 20 : (25 - level);
        return ThreadLocalRandom.current().nextInt(1, bound + 1) == 1;
    }
    @Override
    public void onActivation(Player victim, Player attacker, EntityDamageByEntityEvent ev, int level) {
        attacker.sendMessage("§bLifeSteal §aActivated!");
        int lifeShieldLevel = 0;
        ItemStack chestplate = victim.getInventory().getChestplate();
        if (chestplate != null && chestplate.hasItemMeta()) {
            NamespacedKey key = new NamespacedKey(SkyblockCore.getInstance(), "enchant_" + LIFESHIELD_ID);
            lifeShieldLevel = chestplate.getItemMeta().getPersistentDataContainer().getOrDefault(key, PersistentDataType.INTEGER, 0);
        }
        if (lifeShieldLevel > 0) {
            int blockBound = lifeShieldLevel > 10 ? 3 : 6;
            if (ThreadLocalRandom.current().nextInt(1, blockBound + 1) == 1) {
                attacker.sendMessage("§bLifeSteal §cDeactivated by LifeShield Enchant!");
                return;
            }
        }
        victim.removePotionEffect(PotionEffectType.REGENERATION);
        victim.removePotionEffect(PotionEffectType.HEALTH_BOOST);
        victim.sendMessage("§cStruck by §bLifeSteal §cEnchant! Health Effects removed");
        double healthTaken = level > 9 ? 8 : Math.ceil(level / 2.0);
        victim.setHealth(Math.max(0, victim.getHealth() - healthTaken));
        double maxHealth = attacker.getAttribute(Attribute.MAX_HEALTH).getValue();
        attacker.setHealth(Math.min(maxHealth, attacker.getHealth() + healthTaken));
    }
}
